sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderRule.scala

```diff
@@ -30,4 +30,6 @@ trait CustomShuffleReaderRule extends Rule[SparkPlan] {
    * Returns the list of [[ShuffleOrigin]]s supported by this rule.
    */
   def supportedShuffleOrigins: Seq[ShuffleOrigin]
+
+  def mayAddExtraShuffles: Boolean = false
```

Comment:
I know it's safe. Just in case, a little different behavior is `ValidateRequirements` also checks `outputOrdering`.

Reply:
That means the `CostEvaluator` should also consider the number of sorts, not only shuffles. It's doable, but we need more discussion for this new cost model.